horovod/common/operations.cc

```diff
@@ -267,6 +389,141 @@ const Status DUPLICATE_NAME_ERROR = Status::InvalidArgument(
+    bool op_completed = false;
+
+    std::vector<TensorTableEntry>::size_type i = 0;
+    while (i < tensor_table.size())
+    {
+        op_completed = false;
+        if (tensor_table[i].request == NULL_VAL && tensor_table[i].comm != NULL_VAL)
+        {
+            // Issue op if communicator assigned for this tensor
+            MLSL_LOG(DEBUG, "NOW collective operation %s %p %p %d %p %p %ld\n",
+                     tensor_table[i].tensor_name.c_str(), tensor_table[i].tensor->data(),
+                     tensor_table[i].comm, tensor_table[i].tensor->dtype(), tensor_table[i].request,
+                     NULL_VAL, tensor_table[i].tensor->size());
+
+            if (tensor_table[i].req_type == HRT_BCAST)
+            {
+                tensor_table[i].request = tensor_table[i].comm->Bcast(horovod_global.rank == tensor_table[i].root_rank ?
+                                                   (void*) tensor_table[i].tensor->data() : (void*) tensor_table[i].output->data(),
+                                                   (int) tensor_table[i].tensor->size(), MLSL::DT_BYTE, tensor_table[i].root_rank, MLSL::GT_DATA);
+            }
+            else if (tensor_table[i].req_type == HRT_ALLREDUCE)
+            {
+                tensor_table[i].request = tensor_table[i].comm->AllReduce((void*) tensor_table[i].tensor->data(), (void*) tensor_table[i].output->data(),
+                                                   (int) tensor_table[i].tensor->shape().num_elements(), GetMLSLDataType(tensor_table[i].tensor),
+                                                   MLSL::RT_SUM, MLSL::GT_DATA);
+            }
+            else if (tensor_table[i].req_type == HRT_ALLGATHER && tensor_table[i].size_vec != NULL)
+            {
+                size_t mySize = tensor_table[i].tensor->shape().dim_size(0);
+                std::vector<size_t> rcvCounts(horovod_global.size);
+                for (int j = 0; j < horovod_global.size; j++)
+                    rcvCounts[j] = sizeof(size_t);
+
+                tensor_table[i].request = tensor_table[i].comm->AllGatherv(&mySize, sizeof(size_t), (void*) tensor_table[i].size_vec, rcvCounts.data(),
+                                                                             MLSL::DT_BYTE, MLSL::GT_DATA);
+                tensor_table[i].tmp_comm = tensor_table[i].comm;
+            }
+
+            tensor_table[i].comm = NULL_VAL;
+            MLSL_LOG(DEBUG, "NOW collective operation done %s %p %p %d %p %p %ld\n",
+                     tensor_table[i].tensor_name.c_str(), tensor_table[i].tensor->data(),
+                     tensor_table[i].comm, tensor_table[i].tensor->dtype(), tensor_table[i].request,
+                     NULL_VAL, tensor_table[i].tensor->size());
+        }
+        else if (tensor_table[i].request != NULL_VAL && tensor_table[i].comm == NULL_VAL)
+        {
+            bool testflag = 0;
+            MLSL::Environment::GetEnv().Test(tensor_table[i].request, &testflag);
```

Comment:
What does this test mean, and what if it fails?

Reply:
@SmorkalovME  Can you explain where the `callback` function is executed or why doesn't it need to be executed in the case of: `tensor_table[i].request == NULL_VAL && tensor_table[i].comm != NULL_VAL`

Reply:
@SmorkalovME By the way, Can you show the performance improvement between `horovod allreduce/broadcast/allgather` and `mlsl allreduce/bcast/allgather` ?  thanks.

Reply:
@XiaotaoChen Because here we work with non-blocking collective operations. In this piece of code tensor_table[i].request == NULL_VAL && tensor_table[i].comm != NULL_VAL we start collective operations. Here tensor_table[i].request != NULL_VAL && tensor_table[i].comm == NULL_VAL we check whether the operation was terminated. If it was, then, in case of the called operation being ALLGATHER of the first stage, we call another non-blocking ALLGATHER (the second stage) and wait for its completion, but in other cases we set op_completed = true and call callback function

Reply:
@shirosankaku Thanks for your detailed explanation. got that.  So the main difference is that `mlsl` work with non-blocking collective operations, but the default implementation work with blocking mode. Non-blocking mode can make better use of bandwidth theoretically. Can you tell the performance improvement between the two implementations in some concrete cases.

Reply:
Let's rename all these testflags to `is_completed` (or something along those lines) for clarity.